linera-chain/src/chain.rs

```diff
@@ -672,18 +673,43 @@ where
                     index: message.event.index,
                 },
                 authenticated_signer: message.event.authenticated_signer,
+                refund_grant_to: message.event.refund_grant_to,
             };
             let outcomes = match message.action {
-                MessageAction::Accept => self
-                    .execution_state
-                    .execute_message(
-                        context,
-                        message.event.message.clone(),
-                        &mut resource_controller,
-                    )
-                    .await
-                    .map_err(|err| ChainError::ExecutionError(err, chain_execution_context))?,
+                MessageAction::Accept => {
+                    let mut grant = message.event.grant;
+                    let mut outcomes = self
+                        .execution_state
+                        .execute_message(
+                            context,
+                            message.event.message.clone(),
+                            if grant > Amount::ZERO {
+                                Some(&mut grant)
+                            } else {
+                                None
+                            },
```

Comment:
Maybe more concise (also below):

```suggestion
                            (grant > Amount::ZERO).then_some(&mut grant),
```

Reply:
Sg. It doesn't apply below because the `then`-branch is an option.